hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala

```diff
@@ -0,0 +1,2259 @@
+    *
+    * The stream's element region must be assigned by a consumer before initialize
+    * is called.
+    *
+    * This block cannot jump away, e.g. to `LendOfStream`.
+    *
+    */
+  def initialize(cb: EmitCodeBuilder): Unit
+
+  /**
+    * Stream element region, into which the `element` is emitted. The assignment, clearing,
+    * and freeing of the element region is the responsibility of the stream consumer.
+    */
+  val elementRegion: Settable[Region]
+
+  /**
+    * This boolean parameter indicates whether the producer's elements should be emitted into
+    * separate regions (by clearing when elements leave a consumer's scope). This parameter
+    * propagates bottom-up from producers like [[ReadPartition]] and [[StreamRange]], but
+    * it is the responsibility of consumers to implement the right memory management semantics
+    * based on this flag.
+    */
+  val separateRegions: Boolean
+
+  /**
+    * The `LproduceElement` label is the mechanism by which consumers drive iteration. A consumer
+    * jumps to `LproduceElement` when it is ready for an element. The code block at this label,
+    * defined by the producer, jumps to either `LproduceElementDone` or `LendOfStream`, both of
+    * which the consumer must define.
+    */
+  val LproduceElement: CodeLabel
+
+  /**
+    * The `LproduceElementDone` label is jumped to by the code block at `LproduceElement` if
+    * the stream has produced a valid `element`. The immediate stream consumer must define
+    * this label.
+    */
+  final val LproduceElementDone: CodeLabel = CodeLabel()
+
+  /**
+    * The `LendOfStream` label is jumped to by the code block at `LproduceElement` if
+    * the stream has no more elements to return. The immediate stream consumer must
+    * define this label.
+    */
+  final val LendOfStream: CodeLabel = CodeLabel()
+
+
+  /**
+    * Stream element. This value is valid after the producer jumps to `LproduceElementDone`,
+    * until a consumer jumps to `LproduceElement` again, or calls `close()`.
+    */
+  val element: EmitCode
+
+  /**
+    * Stream producer cleanup method. If `initialize` is called, then the `close` method
+    * must be called as well to properly handle owned resources like files.
+    */
+  def close(cb: EmitCodeBuilder): Unit
+
+  final def unmanagedConsume(cb: EmitCodeBuilder, setup: EmitCodeBuilder => Unit = _ => ())(perElement: EmitCodeBuilder => Unit): Unit = {
+
+    this.initialize(cb)
+    setup(cb)
+    cb.goto(this.LproduceElement)
+    cb.define(this.LproduceElementDone)
+    perElement(cb)
+    cb.goto(this.LproduceElement)
+
+    cb.define(this.LendOfStream)
+    this.close(cb)
+  }
+
+  // only valid if `perElement` does not retain pointers into the element region after it returns (or adds region references)
+  final def memoryManagedConsume(outerRegion: Value[Region], cb: EmitCodeBuilder, setup: EmitCodeBuilder => Unit = _ => ())(perElement: EmitCodeBuilder => Unit): Unit = {
+
+    if (separateRegions)
```

Comment:
If separateRegions is false, can you just fall back to `unmanagedConsume`?

Reply:
good call, done.